File: src/3484.c
#include <ctype.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "uthash/uthash.h"  // https://troydhanson.github.io/uthash/

#ifndef HASH_H
#define HASH_H

typedef struct {
    char* key;
    int val;
    UT_hash_handle hh;
} HashItem;
HashItem* hashFindItem(HashItem** obj, const char* key) {
    HashItem* pEntry = NULL;
    HASH_FIND_STR(*obj, key, pEntry);
    return pEntry;
}
bool hashAddItem(HashItem** obj, char* key, int val) {
    if (hashFindItem(obj, key)) {
        return false;
    }
    HashItem* pEntry = (HashItem*)malloc(sizeof(HashItem));
    pEntry->key = strdup(key);
    pEntry->val = val;
    HASH_ADD_STR(*obj, key, pEntry);
    return true;
}
bool hashSetItem(HashItem** obj, char* key, int val) {
    HashItem* pEntry = hashFindItem(obj, key);
    if (!pEntry) {
        hashAddItem(obj, key, val);
    } else {
        pEntry->val = val;
    }
    return true;
}
int hashGetItem(HashItem** obj, const char* key, int defaultVal) {
    HashItem* pEntry = hashFindItem(obj, key);
    if (!pEntry) {
        return defaultVal;
    }
    return pEntry->val;
}
bool hashErase(HashItem** obj, char* key) {
    HashItem* pEntry = hashFindItem(obj, key);
    if (!pEntry) {
        return false;
    }
    HASH_DEL(*obj, pEntry);
    free(pEntry->key);
    free(pEntry);
    return true;
}
void hashFree(HashItem** obj) {
    HashItem *curr = NULL, *tmp = NULL;
    HASH_ITER(hh, *obj, curr, tmp) {
        HASH_DEL(*obj, curr);
        free(curr->key);
        free(curr);
    }
}

#endif  // HASH_H
typedef struct {
    HashItem* cellValues;
} Spreadsheet;
Spreadsheet* spreadsheetCreate(int rows) {
    Spreadsheet* obj = NULL;

    obj = (Spreadsheet*)malloc(sizeof(Spreadsheet));
    if (obj == NULL) {
        perror("malloc");
        return obj;
    }
    obj->cellValues = NULL;

    return obj;
}
void spreadsheetSetCell(Spreadsheet* obj, char* cell, int value) {
    //
    hashSetItem(&obj->cellValues, cell, value);
}
void spreadsheetResetCell(Spreadsheet* obj, char* cell) {
    //
    hashErase(&obj->cellValues, cell);
}
int getCellVal(Spreadsheet* obj, const char* cell) {
    int retVal = 0;

    if (isalpha(cell[0])) {
        retVal = hashGetItem(&obj->cellValues, cell, 0);
    } else {
        retVal = atoi(cell);
    }

    return retVal;
}
int spreadsheetGetValue(Spreadsheet* obj, char* formula) {
    int retVal = 0;

    char* plus = strchr(formula, '+');

    char* cell1 = (char*)malloc(plus - formula);
    strncpy(cell1, formula + 1, plus - formula - 1);
    cell1[plus - formula - 1] = '\0';
    char* cell2 = strdup(plus + 1);

    retVal = getCellVal(obj, cell1) + getCellVal(obj, cell2);

    free(cell1);
    free(cell2);

    return retVal;
}
void spreadsheetFree(Spreadsheet* obj) {
    hashFree(&obj->cellValues);
    free(obj);
}
/**
 * Your Spreadsheet struct will be instantiated and called as such:
 * Spreadsheet* obj = spreadsheetCreate(rows);
 * spreadsheetSetCell(obj, cell, value);
 * spreadsheetResetCell(obj, cell);
 * int param_3 = spreadsheetGetValue(obj, formula);
 * spreadsheetFree(obj);
 */

int main(int argc, char** argv) {
#define MAX_SIZE (100)
    struct testCaseType {
        char* methods;
        char* cell;
        int value;
        char* formula;
    } testCase[] = {{"Spreadsheet", NULL, 3, NULL}, {"getValue", NULL, 0, "=5+7"},  {"setCell", "A1", 10, NULL},
                    {"getValue", NULL, 0, "=A1+6"}, {"setCell", "B2", 15, NULL},    {"getValue", NULL, 0, "=A1+B2"},
                    {"resetCell", "A1", 0, NULL},   {"getValue", NULL, 0, "=A1+B2"}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input:
     *  ["Spreadsheet", "getValue", "setCell", "getValue", "setCell", "getValue", "resetCell", "getValue"]
     *  [[3], ["=5+7"], ["A1", 10], ["=A1+6"], ["B2", 15], ["=A1+B2"], ["A1"], ["=A1+B2"]]
     *  Output:
     *  [null, 12, null, 16, null, 25, null, 15]
     */

    printf("Input\n");
    printf("[");
    for (int i = 0; i < numberOfTestCase; ++i) {
        printf("%s\"%s\"", (i == 0) ? "" : ", ", testCase[i].methods);
    }
    printf("]\n");
    printf("[");
    for (int i = 0; i < numberOfTestCase; ++i) {
        printf("%s", (i == 0) ? "" : ", ");
        if (strcmp(testCase[i].methods, "Spreadsheet") == 0) {
            printf("[%d]", testCase[i].value);
        } else if (strcmp(testCase[i].methods, "getValue") == 0) {
            printf("[\"%s\"]", testCase[i].formula);
        } else if (strcmp(testCase[i].methods, "setCell") == 0) {
            printf("[\"%s\", %d]", testCase[i].cell, testCase[i].value);
        } else if (strcmp(testCase[i].methods, "resetCell") == 0) {
            printf("[\"%s\"]", testCase[i].cell);
        }
    }
    printf("]\n");

    Spreadsheet* pSpreadsheet = NULL;
    printf("Output\n");
    printf("[");
    for (int i = 0; i < numberOfTestCase; ++i) {
        if (strcmp(testCase[i].methods, "Spreadsheet") == 0) {
            pSpreadsheet = spreadsheetCreate(testCase[i].value);
            printf("%snull", (i == 0) ? "" : ", ");
        } else if (strcmp(testCase[i].methods, "getValue") == 0) {
            int ret = spreadsheetGetValue(pSpreadsheet, testCase[i].formula);
            printf("%s%d", (i == 0) ? "" : ", ", ret);
        } else if (strcmp(testCase[i].methods, "setCell") == 0) {
            spreadsheetSetCell(pSpreadsheet, testCase[i].cell, testCase[i].value);
            printf("%snull", (i == 0) ? "" : ", ");
        } else if (strcmp(testCase[i].methods, "resetCell") == 0) {
            spreadsheetResetCell(pSpreadsheet, testCase[i].cell);
            printf("%snull", (i == 0) ? "" : ", ");
        }
    }
    printf("]\n");
    spreadsheetFree(pSpreadsheet);
    pSpreadsheet = NULL;

    return EXIT_SUCCESS;
}

File: src/3484.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Spreadsheet:
    def __init__(self, rows: int):
        self.cell_values = {}

    def setCell(self, cell: str, value: int) -> None:
        self.cell_values[cell] = value

    def resetCell(self, cell: str) -> None:
        if cell in self.cell_values:
            del self.cell_values[cell]

    def getValue(self, formula: str) -> int:
        retVal = 0

        i = formula.find('+')
        cell1 = formula[1:i]
        cell2 = formula[i + 1:]
        val1 = self.cell_values.get(cell1, 0) if cell1[0].isalpha() else int(cell1)
        val2 = self.cell_values.get(cell2, 0) if cell2[0].isalpha() else int(cell2)
        retVal = val1 + val2

        return retVal

# Your Spreadsheet object will be instantiated and called as such:
# obj = Spreadsheet(rows)
# obj.setCell(cell,value)
# obj.resetCell(cell)
# param_3 = obj.getValue(formula)


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        logging.info("Input")
        for methods, val in zip(
            [["Spreadsheet", "getValue", "setCell", "getValue", "setCell", "getValue", "resetCell", "getValue"]],
            [[[3], ["=5+7"], ["A1", 10], ["=A1+6"], ["B2", 15], ["=A1+B2"], ["A1"], ["=A1+B2"]]]
        ):
            # /* Example
            #  *  Input:
            #  *  ["Spreadsheet", "getValue", "setCell", "getValue", "setCell", "getValue", "resetCell", "getValue"]
            #  *  [[3], ["=5+7"], ["A1", 10], ["=A1+6"], ["B2", 15], ["=A1+B2"], ["A1"], ["=A1+B2"]]
            #  *  Output:
            #  *  [null, 12, null, 16, null, 25, null, 15]
            #  */
            logging.info("%s", methods)
            logging.info("%s", val)

            displayStr = []
            for key, value in zip(methods, val):
                if key == "Spreadsheet":
                    obj = Spreadsheet(value)
                    displayStr.append(None)
                elif key == "getValue":
                    retVal = obj.getValue(value[0])
                    displayStr.append(retVal)
                elif key == "setCell":
                    obj.setCell(value[0], value[1])
                    displayStr.append(None)
                elif key == "resetCell":
                    retVal = obj.resetCell(value[0])
                    displayStr.append(retVal)
            logging.info("Output")
            logging.info("%s", displayStr)

            print("")
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

        print("")
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

        sys.exit(0)
